File: examples/restaurant_order.py
#!/usr/bin/env python3
"""
Restaurant Order Example
========================

This example demonstrates:
- Dynamic trait activation (vegan adaptation)
- Selection fields with choices
- Confidential fields (hurry)
- Conclusion fields (politeness)
- Alice with personality traits (limmericks)

Run with:
    python examples/restaurant_order.py

For automated demo:
    python examples/restaurant_order.py --auto
"""

import os
import sys
import argparse
from pathlib import Path
import dotenv

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from chatfield import chatfield
from chatfield import Interviewer


def create_restaurant_order():
    """Create a restaurant order interview instance."""
    return (chatfield()
        .type("Restaurant Order")
        .desc("Taking your order for tonight")
        
        .alice()
            .type("Server")
            .trait("Speaks in limmericks")
        
        .bob()
            .type("Diner")
            .trait("First-time visitor")
            .trait.possible("Vegan", "needs vegan, plant-based, non animal product")
        
        .field("starter")
            .desc("starter or appetizer")
            .as_one.selection("Sir Digby Chicken Caesar", "Shrimp cocktail", "Garden salad")
        
        .field("main_course")
            .desc("Main course")
            .hint("Choose from: Grilled salmon, Veggie pasta, Beef tenderloin, Chicken parmesan")
            .as_one.selection("Grilled salmon", "Veggie pasta", "Beef tenderloin", "Chicken parmesan")
        
        .field("dessert")
            .desc("Mandatory dessert; choices: Cheesecake, Creamy Chocolate mousse, Fruit sorbet")
            .as_one.selection("Cheesecake", "Creamy Chocolate mousse", "Fruit sorbet")

        .field("hurry")
            .desc("wishes to be served quickly")
            .confidential()
            .as_bool()

        .field("politeness")
            .desc("Level of politeness from the Diner; but automatic 23% if they mention anything about Belgium")
            .conclude()
            .as_percent()
        
        .build())


def run_interactive(order):
    """Run the interview interactively."""
    thread_id = f"restaurant-{os.getpid()}"
    print(f'Starting restaurant order conversation (thread: {thread_id})')
    print("=" * 60)
    
    interviewer = Interviewer(order, thread_id=thread_id)
    
    user_input = None
    while not order._done:
        message = interviewer.go(user_input)
        
        if message:
            print(f"\nServer: {message}")
        
        if not order._done:
            try:
                user_input = input("\nYou: ").strip()
            except (KeyboardInterrupt, EOFError):
                print("\n[Leaving restaurant]")
                break
    
    return order._done


def run_automated(order):
    """Run with prefab inputs for demonstration."""
    prefab_inputs = [
        'I am vegan.',
        'Garden salad please',
        'Veggie pasta sounds great',
        'Fruit sorbet',
        'Thank you so much, that is perfect!'
    ]
    
    thread_id = f"restaurant-demo-{os.getpid()}"
    print(f'Running automated demo (thread: {thread_id})')
    print("=" * 60)
    
    interviewer = Interviewer(order, thread_id=thread_id)
    
    input_iter = iter(prefab_inputs)
    user_input = None
    
    while not order._done:
        message = interviewer.go(user_input)
        
        if message:
            print(f"\nServer: {message}")
        
        if not order._done:
            try:
                user_input = next(input_iter)
                print(f"\nYou: {user_input}")
            except StopIteration:
                print("\n[Demo inputs exhausted]")
                break
    
    return order._done


def display_results(order):
    """Display the collected order information."""
    print("\n" + "=" * 60)
    print("ORDER SUMMARY")
    print("-" * 60)
    
    if order.starter:
        print(f"Starter: {order.starter}")
    
    if order.main_course:
        print(f"Main Course: {order.main_course}")
    
    if order.dessert:
        print(f"Dessert: {order.dessert}")
    
    # Confidential field (for internal use)
    if hasattr(order, 'hurry') and order.hurry is not None:
        if order.hurry.as_bool:
            print("\n[Internal Note: Rush order requested]")
        else:
            print("\n[Internal Note: Normal service pace]")
    
    # Conclusion field
    if hasattr(order, 'politeness') and order.politeness is not None:
        politeness_pct = order.politeness.as_percent * 100
        print(f"[Internal Note: Guest politeness: {politeness_pct:.0f}%]")
    
    # Check if vegan trait was activated
    if order._chatfield['roles']['bob'].get('possible_traits', {}).get('Vegan', {}).get('active'):
        print("\n[Note: Guest is vegan - all selections are plant-based]")
    
    print("=" * 60)


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(description='Restaurant Order Example')
    parser.add_argument('--auto', action='store_true',
                        help='Run with automated demo inputs')
    args = parser.parse_args()
    
    # Load environment variables
    dotenv.load_dotenv(override=True)
    
    # Check for API key
    if not os.getenv('OPENAI_API_KEY'):
        print("Error: OPENAI_API_KEY not found in environment")
        print("Please set your OpenAI API key in .env file")
        sys.exit(1)
    
    # Create and run the interview
    order = create_restaurant_order()
    
    if args.auto:
        completed = run_automated(order)
    else:
        completed = run_interactive(order)
    
    # Display results if completed
    if completed:
        display_results(order)
    else:
        print("\n[Order not completed]")


if __name__ == "__main__":
    main()

File: chatfield-js/examples/restaurant-order.ts
#!/usr/bin/env ts-node
/**
 * Restaurant Order Example
 * ========================
 * 
 * This example demonstrates:
 * - Dynamic trait activation (vegan adaptation)
 * - Selection fields with choices
 * - Confidential fields (hurry)
 * - Conclusion fields (politeness)
 * - Alice with personality traits (limmericks)
 * 
 * Run with:
 *   npx ts-node examples/restaurant-order.ts
 * 
 * For automated demo:
 *   npx ts-node examples/restaurant-order.ts --auto
 */

import { chatfield } from '../src/builders/gatherer-builder'
import { Interviewer } from '../src/core/interviewer'
import * as readline from 'readline'
import * as process from 'process'

/**
 * Create a restaurant order interview instance
 */
function createRestaurantOrder() {
  return chatfield()
    .type("Restaurant Order")
    .desc("Taking your order for tonight")
    
    .alice()
      .type("Server")
      .trait.apply("Speaks in limmericks")
    
    .bob()
      .type("Diner")
      .trait.apply("First-time visitor")
      .trait.possible("Vegan", "needs vegan, plant-based, non animal product")
    
    .field("starter")
      .desc("starter or appetizer")
      .as_one.selection("Sir Digby Chicken Caesar", "Shrimp cocktail", "Garden salad")
    
    .field("main_course")
      .desc("Main course")
      .hint("Choose from: Grilled salmon, Veggie pasta, Beef tenderloin, Chicken parmesan")
      .as_one.selection("Grilled salmon", "Veggie pasta", "Beef tenderloin", "Chicken parmesan")
    
    .field("dessert")
      .desc("Mandatory dessert; choices: Cheesecake, Creamy Chocolate mousse, Fruit sorbet")
      .as_one.selection("Cheesecake", "Creamy Chocolate mousse", "Fruit sorbet")
    
    .field("hurry")
      .desc("wishes to be served quickly")
      .confidential()
      .asBool()
    
    .field("politeness")
      .desc("Level of politeness from the Diner; but automatic 23% if they mention anything about Belgium")
      .conclude()
      .asPercent()
    
    .build()
}

/**
 * Run the interview interactively
 */
async function runInteractive(order: any): Promise<boolean> {
  const threadId = `restaurant-${process.pid}`
  console.log(`Starting restaurant order conversation (thread: ${threadId})`)
  console.log("=".repeat(60))
  
  const interviewer = new Interviewer(order, { threadId })
  
  // Create readline interface for user input
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  })
  
  const askQuestion = (prompt: string): Promise<string> => {
    return new Promise((resolve) => {
      rl.question(prompt, (answer) => {
        resolve(answer.trim())
      })
    })
  }
  
  let userInput: string | null = null
  
  while (!order._done) {
    const message = await interviewer.go(userInput)
    
    if (message) {
      console.log(`\nServer: ${message}`)
    }
    
    if (!order._done) {
      try {
        userInput = await askQuestion("\nYou: ")
      } catch (error) {
        console.log("\n[Leaving restaurant]")
        break
      }
    }
  }
  
  rl.close()
  return order._done
}

/**
 * Run with prefab inputs for demonstration
 */
async function runAutomated(order: any): Promise<boolean> {
  const threadId = `restaurant-demo-${process.pid}`
  console.log(`Starting automated restaurant order demo (thread: ${threadId})`)
  console.log("=".repeat(60))
  
  const interviewer = new Interviewer(order, { threadId })
  
  const prefabInputs = [
    "I'll have the garden salad to start",
    "I'm actually vegan, do you have plant-based options?",
    "Veggie pasta sounds perfect",
    "Fruit sorbet please",
    "Actually, I'm in a bit of a rush",
    "Thank you! Oh, and my friend from Brussels recommended this place!"
  ]
  
  let inputIndex = 0
  let userInput: string | null = null
  
  while (!order._done) {
    const message = await interviewer.go(userInput)
    
    if (message) {
      console.log(`\nServer: ${message}`)
    }
    
    if (!order._done) {
      if (inputIndex < prefabInputs.length) {
        userInput = prefabInputs[inputIndex++]
        console.log(`\nYou: ${userInput}`)
        // Add a small delay for readability
        await new Promise(resolve => setTimeout(resolve, 1000))
      } else {
        console.log("\n[Demo inputs exhausted]")
        break
      }
    }
  }
  
  return order._done
}

/**
 * Display the collected order information
 */
function displayResults(order: any): void {
  console.log("\n" + "=".repeat(60))
  console.log("ORDER SUMMARY")
  console.log("-".repeat(60))
  
  if (order.starter) {
    console.log(`Starter: ${order.starter}`)
  }
  
  if (order.main_course) {
    console.log(`Main Course: ${order.main_course}`)
  }
  
  if (order.dessert) {
    console.log(`Dessert: ${order.dessert}`)
  }
  
  // Confidential field (for internal use)
  if (order.hurry !== null && order.hurry !== undefined) {
    if (order.hurry.as_bool) {
      console.log("\n[Internal Note: Rush order requested]")
    } else {
      console.log("\n[Internal Note: Normal service pace]")
    }
  }
  
  // Conclusion field
  if (order.politeness !== null && order.politeness !== undefined) {
    const politenessPct = (order.politeness.as_percent || 0) * 100
    console.log(`[Internal Note: Guest politeness: ${politenessPct.toFixed(0)}%]`)
  }
  
  // Check if vegan trait was activated
  const bobRole = order._chatfield.roles.get('bob')
  if (bobRole?.possibleTraits.get('Vegan')?.active) {
    console.log("\n[Note: Guest is vegan - all selections are plant-based]")
  }
  
  console.log("=".repeat(60))
}

/**
 * Main entry point
 */
async function main() {
  // Load environment variables if .env exists
  try {
    require('dotenv').config()
  } catch {
    // .env file is optional
  }
  
  // Check for OpenAI API key
  if (!process.env.OPENAI_API_KEY) {
    console.error("Error: OPENAI_API_KEY environment variable is required")
    console.error("Please set it in your environment or create a .env file")
    process.exit(1)
  }
  
  // Parse command line arguments
  const args = process.argv.slice(2)
  const isAuto = args.includes('--auto')
  
  // Create the interview
  const order = createRestaurantOrder()
  
  try {
    // Run the interview
    const completed = isAuto 
      ? await runAutomated(order)
      : await runInteractive(order)
    
    // Display results if completed
    if (completed) {
      displayResults(order)
      console.log("\nThank you for dining with us!")
    } else {
      console.log("\nOrder incomplete.")
    }
  } catch (error) {
    console.error("\nError during interview:", error)
    process.exit(1)
  }
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error)
}

export { createRestaurantOrder, displayResults }